pgwire: decorate TransactionRetryWithProtoRefreshError with docs URL

otan opened the pull request:
Release note (sql change): TransactionRetryWithProtoRefreshError now
have a HINT field that links to a web page containing useful
information to resolve the error.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/55197)
<!-- Reviewable:end -->

awoods187 commented:
This is awesome! I'd love to backport this to 20.2.0.

Will this work for all of the retry errors? https://www.cockroachlabs.com/docs/v20.1/transaction-retry-error-reference.html

otan commented:
Yeah, I think it's safe to backport to v20.1 as well.

otan commented:
___
*[pkg/sql/pgwire/conn.go, line 1248 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/55197#-MIgKTmu-06nVqOps_7T:-MIgKTmu-06nVqOps_7U:bax8oeb) ([raw file](https://github.com/cockroachdb/cockroach/blob/314b955d3b285c4f011d0a827fed8ccab9a6344f/pkg/sql/pgwire/conn.go#L1248)):*
> ```Go
> ) error {
> 	// Now send the error to the client, appending useful docs links if applicable.
> 	pgErr := pgerror.Flatten(pgerrordoc.WithDocsDetails(err))
> ```

(ideally the Details live on the generation of the error message, but that seems hard to pass down. I'd like to do this in `Flatten` but that introduces a circular dependency with roachpb and base unless we inject the function in at all call sites. I see we do something similar to this for the PGCode 40001)


<!-- Sent from Reviewable.io -->

rafiss commented on pkg/sql/pgwire/pgerrordoc/pgerrordoc.go:
```diff
@@ -0,0 +1,58 @@
+			}
+		}
+
+		url := base.DocsURL("transaction-retry-error-reference.html")
```
i don't suppose this could make use of our link shortener (so that we can track clicks)?

otan replied:
(i can do this if someone sets it up, which I think should be a separate PR. We can make the `errors.WithIssueLink` do this in general, me thinks.)

rafiss replied:
Separate PR sounds good. it is set up here, btw: https://go.crdb.dev/

(though Bob still has to do something manual to make it exposed publicly)

otan commented:
> i don't suppose this could make use of our link shortener (so that we can track clicks)?

I'll need someone to set that up.

knz commented on pkg/sql/pgwire/pgerrordoc/pgerrordoc.go:
```diff
@@ -0,0 +1,58 @@
+
+		// TransactionRetryWithProtoRefreshError does not actually wrap the error message,
+		// so we're forced to read the message string.
+		for _, match := range []string{
```
Maybe you can use `roachpb.TransactionRetryReason_name` or `roachpb.TransactionRetryReason_value` instad of reconstructing a slice of names again and again here?

(there's also `TransactionAbortedReason_value` / `TransactionAbortedReason_name` to play with)

otan replied:
I chose this initially because i wasn't sure all anchors existed on the docs page, but i guess if the anchor doesn't exist that's ok!

knz commented on pkg/sql/pgwire/pgerrordoc/pgerrordoc.go:
```diff
@@ -0,0 +1,58 @@
+		if pageAnchor != "" {
+			url += "#" + pageAnchor
+		}
+		err = errors.WithDetailf(
```
Why do you not use `errors.WithIssueLink`? 

The word "Issue" in that API call is a misnomer, it's not specific to issues. It gives you a placeholder for a URL an an arbitrary string.

knz reviewed (commented):
> Reviewed 4 of 4 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55197)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan and @rafiss)

___
*[pkg/sql/pgwire/conn.go, line 1248 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/55197#-MIgKTmu-06nVqOps_7T:-MIxvtuIAuVuSLXdySPl:b5b14g3) ([raw file](https://github.com/cockroachdb/cockroach/blob/314b955d3b285c4f011d0a827fed8ccab9a6344f/pkg/sql/pgwire/conn.go#L1248)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

(ideally the Details live on the generation of the error message, but that seems hard to pass down. I'd like to do this in `Flatten` but that introduces a circular dependency with roachpb and base unless we inject the function in at all call sites. I see we do something similar to this for the PGCode 40001)
</blockquote></details>

I was about to say - I re-reviewed the PR today and I am not seeing why you need a separate `pgerrordoc` package and a separate funcion `WithRelevantLinks`.

IMHO the hint should be injected directly in `Flatten`, by modifuing Flatten as follows:

``` go
  // NEW:
  if  resErr.Code == pgcode.SerializationFailure {
    re := regexp.MustCompile(`((?:ABORT_|RETRY_)[A-Z_]*)`)
     reason :=  re.ReplaceAllString(resErr.Message, "$1")
     reason = strings.ReplaceAll(strings.ToLower(reason), "_", "")
     // Add the hint so it gets picked up by FlattenHints below.
     err = errors.WithIssueLink(err,  base.DocsURL("transaction-retry-error-reference.html#" + reason)
  }
  // Populate the details and hints.
  resErr.Hint = errors.FlattenHints(err)
```


<!-- Sent from Reviewable.io -->

otan committed:
base: refactor DocsURL to its own package

Release note: None

otan commented:
___
*[pkg/sql/pgwire/conn.go, line 1248 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/55197#-MIgKTmu-06nVqOps_7T:-MIyWLhbCiCUAsN_JwVJ:b-7cybus) ([raw file](https://github.com/cockroachdb/cockroach/blob/314b955d3b285c4f011d0a827fed8ccab9a6344f/pkg/sql/pgwire/conn.go#L1248)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

I was about to say - I re-reviewed the PR today and I am not seeing why you need a separate `pgerrordoc` package and a separate funcion `WithRelevantLinks`.

IMHO the hint should be injected directly in `Flatten`, by modifuing Flatten as follows:

``` go
  // NEW:
  if  resErr.Code == pgcode.SerializationFailure {
    re := regexp.MustCompile(`((?:ABORT_|RETRY_)[A-Z_]*)`)
     reason :=  re.ReplaceAllString(resErr.Message, "$1")
     reason = strings.ReplaceAll(strings.ToLower(reason), "_", "")
     // Add the hint so it gets picked up by FlattenHints below.
     err = errors.WithIssueLink(err,  base.DocsURL("transaction-retry-error-reference.html#" + reason)
  }
  // Populate the details and hints.
  resErr.Hint = errors.FlattenHints(err)
```
</blockquote></details>

this is done, but had to migrate the base.DocsURL to its own package.


<!-- Sent from Reviewable.io -->

otan committed:
pgwire: decorate TransactionRetryWithProtoRefreshError with docs URL

Release note (sql change): TransactionRetryWithProtoRefreshError now
have a HINT field that links to a web page containing useful
information to resolve the error.

knz reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 8 of 12 files at r3, 5 of 5 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55197#-:-MJ1Kwhy1SbPTQyJhzAL:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @adityamaru and @otan)



<!-- Sent from Reviewable.io -->

otan commented:
TFTR! 

bors r=knz

i'll probably look at backporting this next point release.
